text/0006-bingings.md

```diff
@@ -0,0 +1,584 @@
+
+###### Returns
+* Index of the new promise that uniquely identifies it within the current execution of the method.
+
+---
+```rust
+promise_results_count() -> u64
+```
+If the current function is invoked by a callback we can access the execution results of the promises that
+caused the callback. This function returns the number of complete and incomplete callbacks.
+
+Note, we are only going to have incomplete callbacks once we have `promise_or` combinator.
+###### Normal execution
+* If there is only one callback `promise_results_count()` returns `1`;
+* If there are multiple callbacks (e.g. created through `promise_and`) `promise_results_count()` returns their number.
+* If the function was called not through the callback `promise_results_count()` returns `0`.
+
+
+---
+```rust
+promise_result(result_idx: u64, register_id: u64) -> u64
+```
+If the current function is invoked by a callback we can access the execution results of the promises that
+caused the callback. This function returns the result in blob format and places it into the register.
+
+###### Normal execution
+* If promise result is complete and successful copies its blob into the register;
+* If promise result is complete and failed or incomplete keeps register unused;
+
+###### Returns
+* If promise result is not complete returns `0`;
+* If promise result is complete and successful returns `1`;
+* If promise result is complete and failed returns `2`.
+
+###### Panics
+* If `result_idx` does not correspond to an existing result panics with `InvalidResultIndex`.
+* If copying the blob exhausts the memory limit it panics with `MemoryAccessViolation`.
+
+###### Current bugs
+* We currently have two separate functions to check for result completion and copy it.
+
+---
+```rust
+promise_return(promise_idx: u64)
+```
+When promise `promise_idx` finishes executing its result is considered to be the result of the current function.
+
+###### Panics
+* If `promise_idx` does not correspond to an existing promise panics with `InvalidPromiseIndex`.
+
+###### Current bugs
+* The current name `return_promise` is inconsistent with the naming convention of Promise API.
+
+## Miscellaneous API
+```rust
+value_return(value_len: u64, value_ptr: u64)
+```
+Sets the blob of data as the return value of the contract.
+
+##### Panics
+* If `value_len + value_ptr` exceeds the memory container or points to an unused register it panics with `MemoryAccessViolation`;
+
+---
+```rust
+panic()
+```
+Terminates the execution of the program with panic `GuestPanic`.
+
+---
+```rust
+log_utf8(len: u64, ptr: u64)
+```
+Logs the UTF-8 encoded string. See https://stackoverflow.com/a/5923961 that explains
+that null termination is not defined through encoding.
+
+###### Normal behavior
+If `len == u64::MAX` then treats the string as null-terminated with character `'\0'`;
+
+###### Panics
+* If string extends outside the memory of the guest with `MemoryAccessViolation`;
```

Comment:
Late for a party..

Right now, if contract passes invalid utf8 to log it will mean that whole transaction will became invalid. But isn't it just a matter of a client interpretation? I think the log interpretation (as well as we args) must orthogonal to the virtual machine.